text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,266 @@
+
+State that is contained outside of a LWC (e.g. in a store of some kind) will need to be serialized manually during SSR and rehydrated prior to invocation of `hydrateElement` on the client.
+
+### Bailing out
+
+There is a spectrum of choices possible when encountering an unexpected mismatch during rehydration. The following is a sampling from that spectrum:
+
+1. Bail out of rehydration entirely when a mismatch is detected, throw away the original HTML, and force a full re-render.
+2. Handle easy mismatches (text nodes, certain attrs, etc.) falling back to number 1 for anything more extreme.
+3. Bail out of rehydration for subtrees, rooted at a mismatch.
+4. Bail out of rehydration for subtrees, rooted at the component where the mismatch occurred.
+5. Try to repair as much of the SSR-derived DOM as possible, throwing nothing away explicitly.
+6. Assume no mismatches exist, and break silently.
+
+The approach that we propose adoping is #4. Specifically:
+
+- We detect a mismatch
+- We patch "easily recoverable" mismatches in the DOM, e.g. text-node mismatches
+- For mismatches that are not easily recoverable
+  + We determine the closest LWC ancestor to that mismatch (anywhere from the root component in a SPA to a leaf component for a small widget).
+  + We invalidate the ancestor's subtree, replacing SSR-derived DOM with CSR DOM for that component subtree only.
+
+### Recovering from lifecycle errors
+
+It might happen that an error is thrown while invoking the `connectedCallback` or other lifecycle method. This situation could occur because:
+
+1. The userland lifecycle method has a bug.
+2. The state extracted from the SSR DOM is incorrect for some reason, leading to unexpected inputs and behavior in the lifecycle method.
+
+We will want to recover from #2. And because it won't be possible to differentiate #1 and #2 at runtime, a catch-all solution must be adopted.
+
+It is proposed that, should an error be thrown during `connectedCallback`, we should assume that an unexpected mismatch has occurred, throw away the SSR DOM, and replace it with CSR VDOM. The `connectedCallback` will then be invoked again (`connectedCallback` can fire more than once), this time for the CSR DOM.
+
+If an error occurs again, we can assume its proximate cause is in userland, not LWC.
+
+
+### `hydrateElement`
+
+In a typical LWC web application, a developer might mount their app using the following pattern:
+
+```javascript
+import { createElement } from 'lwc';
+import MyApp from './my-app';
+
+document
+  .querySelector("#root")
+  .appendChild(createElement('my-app', { is: MyApp }));
+```
+
+After the proposed change, if a developer wants to take advantage of rehydration, a developer would instead do:
+
+```javascript
+import { hydrateElement } from 'lwc';
+import MyApp from './my-app';
+
+hydrateElement(MyApp, document.querySelector('#root')); 
+```
+
+### Constructable stylesheets

```

Comment:
IMHO we should have waited to introduce constructible stylesheets (as we were hiding the internal anyway) if this is creating a path of resistance for hydration. But I guess "Everything is obvious once you know the answer"